### PR TITLE
Skip mount point checks when possible during mount cleanup.

### PR DESCRIPTION
**Pull request by cartermckinnon**
#### What type of PR is this?

/kind bug
/kind api-change

#### What this PR does / why we need it:

This is a continuation of https://github.com/kubernetes/kubernetes/pull/109117. Please see that PR for more background.

Calls to `mounter.Unmount` are preceded and followed by expensive mount point checks. These checks are not necessary on *nix-s with a `umount` implementation that performs a similar check itself. This PR adds a mechanism to detect the "safe" behavior, and avoid mount point checks when possible.

This change represents a significant optimization of `CleanupMountPoint`; enabling use-cases where pods have many mounts, and there is high pod churn. We (EKS) have observed several cases of instability and poor node health in such scenarios, which were resolved with this change.

#### Which issue(s) this PR fixes:

No issue available.

#### Special notes for your reviewer:

I chose to add a function to the `Mounter` interface in order to keep the `CleanupMountPoint` implementation generic for Unix and Windows. If the "safe" `umount` behavior is not detected, the existing code paths are unchanged.

#### Does this PR introduce a user-facing change?

```release-note
On compatible systems, a mounter's Unmount implementation is changed to not return an error when the specified target can be detected as not a mount point. On Linux, the behavior of detecting a mount point depends on `umount` command is validated when the mounter is created. Additionally, mount point checks will be skipped in CleanupMountPoint/CleanupMountWithForce if the mounter's Unmount having the changed behavior of not returning error when target is not a mount point.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: staging/src/k8s.io/mount-utils/mount_windows.go
Comment (cartermckinnon) [inline staging/src/k8s.io/mount-utils/mount_windows.go:249]: I used `false` here to avoid changing any behavior on Windows; but looking at the `Unmount` implementation, I think this might be fine to return `true`?

File: staging/src/k8s.io/mount-utils/mount_linux.go
Comment (wongma7) [inline staging/src/k8s.io/mount-utils/mount_linux.go:342]: is this in the right place? if umount exits 1 I think err will be non-nil so this belongs in the above if statement. Will need to test.

1. create Pod with a volume
2. ssh onto Node and unmount one of its volumes
3. delete Pod
4. does this condition actually get hit?

File: staging/src/k8s.io/mount-utils/mount_windows.go
Comment (wongma7) [reply staging/src/k8s.io/mount-utils/mount_windows.go:249]: ya I see the same, it won't make a difference since windows `mounter.List` does nothing and has a todo  https://github.com/kubernetes/kubernetes/blob/9c151bb31a639663a1d79b4684b5e0a0a0488a2d/staging/src/k8s.io/mount-utils/mount_windows.go#L229 . i would leave it false, it is up to whoever eventually implements that `List` todo to figure out when/if CanSafelySkipMountPointCheck should ever return true

File: staging/src/k8s.io/mount-utils/mount_linux.go
Comment (cartermckinnon) [reply staging/src/k8s.io/mount-utils/mount_linux.go:340]: This is within the `if err != nil {}` block; and I think it should come *after* the `ErrNoChildProcess` check (as is). I'll do that test, to confirm.

**Comment by cartermckinnon** [comment]
I'm not able to reproduce this test failure using:
```
go test -run ^Test_Run_OneVolumeDetachFailNodeWithReadWriteOnce$ k8s.io/kubernetes/pkg/controller/volume/attachdetach/reconciler
```

The line immediately preceding the failed assertion is a sleep statement: https://github.com/cartermckinnon/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler_test.go#L655

I'm going to try it again, but this seems like flakiness that isn't related to the changes in this PR.

/retest

File: staging/src/k8s.io/mount-utils/fake_mounter.go
Comment (wongma7) [inline staging/src/k8s.io/mount-utils/fake_mounter.go:217]: can you make this a property of FakeMounter and add a test case to https://github.com/kubernetes/kubernetes/blob/6b44c0debbd43830c19d8e4d81f370a920ee98ed/staging/src/k8s.io/mount-utils/mount_helper_test.go#L29 ?

Then if this is true, we can have umount return "not mounted", and still the directory should be removed at the end of the testcase.

Comment (cartermckinnon) [reply staging/src/k8s.io/mount-utils/fake_mounter.go]: I added a case to `TestDoCleanupMountPoint` that ensures the mount path is deleted when `CanSafelySkipMountPointCheck` is true.

I also added `TestDetectSafeNotMountedBehavior` in `mount_linux_test.go` that mocks out `umount` and ensures the detection is `true` only when the "not mounted" output is present along with an exit code != 0.

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (wongma7) [inline staging/src/k8s.io/mount-utils/mount_helper_common.go:149]: this could be noisy, maybe V(4) to match the message "%q is a mountpoint, unmounting"?

**Review by wongma7 — APPROVED** [review]

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (manugupt1) [inline staging/src/k8s.io/mount-utils/mount_helper_common.go:56]: I am curious; why do you not need to send the mountPath here?

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (cartermckinnon) [reply staging/src/k8s.io/mount-utils/mount_helper_common.go:149]: Ah yes, I'll fix that (it was definitely noisy in my tests)

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (jingxu97) [inline staging/src/k8s.io/mount-utils/mount_helper_common.go:108]: if mountPath is not a mount point, will it return error at line 104 during Unmount? In this case, it will not call removePath in the following, right?

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (mkimuram) [reply staging/src/k8s.io/mount-utils/mount_helper_common.go:56]: In my understanding, whether the `mounter.CanSafelySkipMountPointCheck()` returns true is decided by [`detectSafeNotMountedBehaviorWithExec`](https://github.com/kubernetes/kubernetes/pull/109676/files/7e674280561f4bbc34f48a3fda2281d2978fbf36#diff-5a53703b1baddb2caa650077f26075c59361575c84387e1e9ad5570a738dd473R240-R255). And `detectSafeNotMountedBehaviorWithExec` seems to decide if it can safely skip mount-point check by checking if umount command to temp directory (/tmp/kubelet-detect-safe-umount) shows "not mounted".

Then, the question is "Are umount commands used in a system always the same for all the paths?"  As "EXTERNAL HELPERS" section of [umount(8) man page](https://man7.org/linux/man-pages/man8/umount.8.html) describes, it seems that external helpers like `umount.fuse` may be called. Therefore, just checking temp directory may not always return the behaviors of the umount command for all the path in the system.

Comment (mkimuram) [reply staging/src/k8s.io/mount-utils/mount_helper_common.go:56]: After thinking about it a bit more, I started to think that the umount command that is used when nothing actually mounted on the path may be the same (If helper is decided by the type of the mounted filesystem, nothing mounted would mean no helper). "not mounted" message will only appears in such a situation, so we may be able to rely on this behavior and regard it as mountPath agnostic.

I'm not 100% sure about it, so we will need to check the logic behind how helper is decided and check if there is no corner case.

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (cartermckinnon) [reply staging/src/k8s.io/mount-utils/mount_helper_common.go:108]: It will not return an error. The contract of `CanSafelySkipMountPointChecks` is that `Unmount` does not produce such errors: https://github.com/kubernetes/kubernetes/blob/7e674280561f4bbc34f48a3fda2281d2978fbf36/staging/src/k8s.io/mount-utils/mount.go#L69

The linux impl ignores this error here, if it has detected the 'safe' behavior: https://github.com/kubernetes/kubernetes/blob/7e674280561f4bbc34f48a3fda2281d2978fbf36/staging/src/k8s.io/mount-utils/mount_linux.go#L345

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (cartermckinnon) [reply staging/src/k8s.io/mount-utils/mount_helper_common.go:56]: I'll add the `-i` flag to the `umount` call in `detectSafeNotMountedBehavior`:
```
       -i, --internal-only
           Do not call the /sbin/umount.filesystem helper even if it
           exists. By default such a helper program is called if it
           exists.
```

Which should be safe for the detection part.

For systems with helpers present, if the detection returns `true`, and later a mount path is passed to `CleanupMountPoint` that triggers a helper which somehow does not conform to the expected behavior, it will result in an error. The path will not be removed. I'm not sure whether this issue will surface in a kubelet event or not, but it seems like an acceptable (non-destructive) failure mode for this edge case.

Comment (cartermckinnon) [reply staging/src/k8s.io/mount-utils/mount_helper_common.go:56]: My tests show that the filesystem is resolved even when the target is not a mount point, when `umount` is executed by the root user.

My system has a single `ext4` filesystem at `/`; I created a helper at `/sbin/umount.ext4`:
```sh
#!/bin/sh

echo "ext4 helper reporting for duty! target: $1"
```

I see the desired `not mounted` output as a **non-root user**:
```
% LIBMOUNT_DEBUG=all umount $(mktemp -d) 
30067: libmount:     INIT: library debug mask: 0xffff
30067: libmount:     INIT: library version: 2.30.2
30067: libmount:     INIT:     feature: selinux
30067: libmount:     INIT:     feature: btrfs
30067: libmount:     INIT:     feature: assert
30067: libmount:     INIT:     feature: debug
Available "LIBMOUNT_DEBUG=<name>[,...]|<mask>" debug masks:
   all      [0xffff] : info about all subsystems
   cache    [0x0004] : paths and tags cache
   cxt      [0x0200] : library context (handler)
   diff     [0x0400] : mountinfo changes tracking
   fs       [0x0040] : FS abstraction
   help     [0x0001] : this help
   locks    [0x0010] : mtab and utab locking
   loop     [0x2000] : loop devices routines
   options  [0x0008] : mount options parsing
   tab      [0x0020] : fstab, mtab, mountinfo routines
   update   [0x0080] : mtab, utab updates
   utils    [0x0100] : misc library utils
   monitor  [0x0800] : mount tables monitor
   btrfs    [0x1000] : btrfs specific routines
30067: libmount:      CXT: [0x55d2c4cf18e0]: ----> allocate [RESTRICTED]
30067: libmount:      CXT: [0x55d2c4cf18e0]: umount: /tmp/tmp.NoOgpNRbgU
30067: libmount:      CXT: [0x55d2c4cf18e0]: umount: lookup FS for '/tmp/tmp.NoOgpNRbgU'
30067: libmount:      CXT: [0x55d2c4cf18e0]: checking for writable tab files
30067: libmount:    UTILS: utab: /run/mount/utab
30067: libmount:    CACHE: [0x55d2c4cf2ad0]: alloc
30067: libmount:    CACHE: [0x55d2c4cf2ad0]: canonicalize path /tmp/tmp.NoOgpNRbgU
30067: libmount:    CACHE: [0x55d2c4cf2ad0]: add entry [ 1] (path): /tmp/tmp.NoOgpNRbgU: /tmp/tmp.NoOgpNRbgU
30067: libmount:      CXT: [0x55d2c4cf18e0]: tabfilter ENABLED!
30067: libmount:      TAB: [0x55d2c4cf4730]: alloc
30067: libmount:      TAB: [0x55d2c4cf4730]: set table parser filter
30067: libmount:      TAB: [0x55d2c4cf4730]: mtab parse: #1 read mountinfo
30067: libmount:      TAB: [0x55d2c4cf4730]: /proc/self/mountinfo: start parsing [entries=0, filter=yes]
30067: libmount:       FS: [0x55d2c4cf4790]: free [refcount=0]
30067: libmount:      TAB: [0x55d2c4cf4730]: /proc/self/mountinfo: stop parsing (0 entries)
30067: libmount:      TAB: [0x55d2c4cf4730]: parsing done [filename=/proc/self/mountinfo, rc=0]
30067: libmount:      CXT: [0x55d2c4cf18e0]: mtab requested [nents=0]
30067: libmount:      TAB: [0x55d2c4cf4730]: unset table parser filter
30067: libmount:      CXT: [0x55d2c4cf18e0]: tabfilter disabled
30067: libmount:      CXT: [0x55d2c4cf18e0]: umount: mtab empty
30067: libmount:      CXT: [0x55d2c4cf18e0]: umount: cannot find '/tmp/tmp.NoOgpNRbgU' in mtab
30067: libmount:      CXT: [0x55d2c4cf18e0]: merging mount flags
30067: libmount:      CXT: [0x55d2c4cf18e0]: final flags: VFS=00000000 user=00000000
30067: libmount:      CXT: [0x55d2c4cf18e0]: umount: evaluating permissions
30067: libmount:      CXT: [0x55d2c4cf18e0]: cannot find /tmp/tmp.NoOgpNRbgU in mtab and you are not root
30067: libmount:      CXT: [0x55d2c4cf18e0]: umount is not allowed for you
30067: libmount:      CXT: [0x55d2c4cf18e0]: umount: preparing failed
30067: libmount:      CXT: [0x55d2c4cf18e0]: excode: rc=1 message="not mounted"
umount: /tmp/tmp.NoOgpNRbgU: not mounted.
30067: libmount:      CXT: [0x55d2c4cf18e0]: <---- reset [status=0] ---->
30067: libmount:       FS: [0x55d2c4cf29f0]: free [refcount=0]
30067: libmount:      TAB: [0x55d2c4cf4730]: reset
30067: libmount:      TAB: [0x55d2c4cf4730]: free [refcount=0]
30067: libmount:      CXT: [0x55d2c4cf18e0]: <---- reset [status=0] ---->
30067: libmount:    CACHE: [0x55d2c4cf2ad0]: free [refcount=0]
30067: libmount:      CXT: [0x55d2c4cf18e0]: <---- free
```

However, as a **root user**, the filesystem is determined to be `ext4` and the helper is called:
```
% sudo LIBMOUNT_DEBUG=all umount $(mktemp -d)
28544: libmount:     INIT: library debug mask: 0xffff
28544: libmount:     INIT: library version: 2.30.2
28544: libmount:     INIT:     feature: selinux
28544: libmount:     INIT:     feature: btrfs
28544: libmount:     INIT:     feature: assert
28544: libmount:     INIT:     feature: debug
Available "LIBMOUNT_DEBUG=<name>[,...]|<mask>" debug masks:
   all      [0xffff] : info about all subsystems
   cache    [0x0004] : paths and tags cache
   cxt      [0x0200] : library context (handler)
   diff     [0x0400] : mountinfo changes tracking
   fs       [0x0040] : FS abstraction
   help     [0x0001] : this help
   locks    [0x0010] : mtab and utab locking
   loop     [0x2000] : loop devices routines
   options  [0x0008] : mount options parsing
   tab      [0x0020] : fstab, mtab, mountinfo routines
   update   [0x0080] : mtab, utab updates
   utils    [0x0100] : misc library utils
   monitor  [0x0800] : mount tables monitor
   btrfs    [0x1000] : btrfs specific routines
28544: libmount:      CXT: [0x55abb03ea8e0]: ----> allocate 
28544: libmount:      CXT: [0x55abb03ea8e0]: umount: /tmp/tmp.TBIZ7scVSe
28544: libmount:      CXT: [0x55abb03ea8e0]: checking for writable tab files
28544: libmount:    UTILS: utab: /run/mount/utab
28544: libmount:      CXT: [0x55abb03ea8e0]: enabling flag 0002
28544: libmount:      CXT: [0x55abb03ea8e0]: umount: mountinfo unnecessary [type=ext4]
28544: libmount:      CXT: [0x55abb03ea8e0]: merging mount flags
28544: libmount:      CXT: [0x55abb03ea8e0]: final flags: VFS=00000000 user=00000000
28544: libmount:      CXT: [0x55abb03ea8e0]: /sbin/umount.ext4         ... found
28544: libmount:      CXT: [0x55abb03ea8e0]: prepare update
28544: libmount:      CXT: [0x55abb03ea8e0]: skip update: NOMTAB flag
28545: libmount:      CXT: [0x55abb03ea8e0]: argv[0] = "/sbin/umount.ext4"
28545: libmount:      CXT: [0x55abb03ea8e0]: argv[1] = "/tmp/tmp.TBIZ7scVSe"
28545: libmount:      CXT: [0x55abb03ea8e0]: argv[2] = "-n"
ext4 helper reporting for duty! target: /tmp/tmp.TBIZ7scVSe
28544: libmount:      CXT: [0x55abb03ea8e0]: /sbin/umount.ext4 executed [status=0]
28544: libmount:      CXT: [0x55abb03ea8e0]: don't update: NOMTAB flag
28544: libmount:      CXT: [0x55abb03ea8e0]: excode: rc=0 message=""
28544: libmount:      CXT: [0x55abb03ea8e0]: <---- reset [status=1] ---->
28544: libmount:       FS: [0x55abb03ea9e0]: free [refcount=0]
28544: libmount:      CXT: [0x55abb03ea8e0]: <---- reset [status=0] ---->
28544: libmount:      CXT: [0x55abb03ea8e0]: <---- free
```

I'm still trying to understand why `ext4` is decided as the filesystem type, this seems to happen in this call (or the next line): https://github.com/util-linux/util-linux/blob/master/libmount/src/context_umount.c#L281

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (manugupt1) [reply staging/src/k8s.io/mount-utils/mount_helper_common.go:108]: Q: Why do we need this call here? Since `mounter.withSafeNotMountedBehavior && strings.Contains(string(output), errNotMounted)` already hides the error. can you not ignore this check?

File: staging/src/k8s.io/mount-utils/mount.go
Comment (manugupt1) [inline staging/src/k8s.io/mount-utils/mount.go:72]: can this be a package private function? Not sure if external consumers should be able to consume it.

Comment (cartermckinnon) [reply staging/src/k8s.io/mount-utils/mount.go:72]: I would tend to agree; this looks a lot like an implementation detail that probably isn't very useful or straightforward to use by a library consumer.

But, because this interface exposes `IsLikelyMountPoint`, I think it's important to expose this function as well. IMO that function should also probably not be exposed, but I think that's outside the scope of this PR.

Comment (manugupt1) [reply staging/src/k8s.io/mount-utils/mount.go:72]: IsLikelyNotMountPoint has a specific use-case where we can detect mount points faster and avoid reading /proc/mounts. So, if you are 100% sure; that a given path is not a bind mount, IsLikelyNotMountPoint should work, it is also used in other places in the use cases.

I think we should keep it package private. Also, if there is a use-case in the future, making it public should be easier than making it private.

Comment (cartermckinnon) [reply staging/src/k8s.io/mount-utils/mount.go:72]: My point was that `IsLikelyNotMountPoint` shouldn't be called in certain circumstances, based on the result of `CanSafelySkipMountPointChecks`, as in `CleanupMountPoints`. I don't feel strongly either way about making it private, so I'll make it private. I agree that it's easy to make it public later if need be.

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (yuefido) [reply staging/src/k8s.io/mount-utils/mount_helper_common.go:56]: Yes. The umount. suffix will be called if there is one under /sbin/ . I don't know for ext4. But for nfs and nfs4, umount.nfs or umount.nfs4 will all check if the mount point is in the mtab and is nfs type file system (https://github.com/stefanha/nfs-utils/blob/vsock/utils/mount/nfsumount.c#L284)

Comment (cartermckinnon) [reply staging/src/k8s.io/mount-utils/mount_helper_common.go]: FYI, @yuefido is on the EFS team -- I asked him to take a look at this because he has additional context on how this PR impacts our (EKS) users.

**Comment by cartermckinnon** [comment]
@jingxu97 @manugupt1 are there any outstanding changes you need on this?

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (manugupt1) [inline staging/src/k8s.io/mount-utils/mount_helper_common.go:151]: can you please update the comment when can errNotExist be returned?

Comment (manugupt1) [reply staging/src/k8s.io/mount-utils/mount_helper_common.go:151]: Recommendation now is to use: errors.Is(err, fs.ErrNotExist) https://pkg.go.dev/os#IsNotExist

File: staging/src/k8s.io/mount-utils/mount_linux.go
Comment (manugupt1) [inline staging/src/k8s.io/mount-utils/mount_linux.go:242]: This is a library and we should not have logging; can we only have debug logs as there a lot of cases where it might be useful?

**Review by manugupt1 — APPROVED** [review]
Small nits; but I am good overall.

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (jingxu97) [inline staging/src/k8s.io/mount-utils/mount_helper_common.go:56]: you only add the logic only to CleanupMountWithForce, which I only see nfs plugin uses it. So this skip logic only for nfs plugin?

File: staging/src/k8s.io/mount-utils/mount_linux.go
Comment (jingxu97) [inline staging/src/k8s.io/mount-utils/mount_linux.go:74]: each time a new mounter is created, this function detectSafeNotMountedBehavior will be called, right? Is there any performance concern?

Comment (cartermckinnon) [reply staging/src/k8s.io/mount-utils/mount_linux.go:74]: My understanding is that the mounter *is* only created once at kubelet boot, here: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubelet/app/server.go#L366

I don't think there's a performance concern unless a mounter is being created frequently, like once a second. I'm not aware of any code paths that do such a thing. I don't think the performance characteristics of `detectSafeNotMountedBehavior` differ significantly from the existing `detectSystemd`.

Comment (jingxu97) [reply staging/src/k8s.io/mount-utils/mount_linux.go:74]: got it, then it should be good.

Comment (jingxu97) [reply staging/src/k8s.io/mount-utils/mount_linux.go:74]: btw: do you know how linux "umount" command works to check whether it is a mount point. It should also check mount table, right? Is it much more efficient?

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (jingxu97) [inline staging/src/k8s.io/mount-utils/mount_helper_common.go:56]: to make it easy to reason the code, I kind of feel we could reduce the number of times to check canSafelySkipMountPointCheck, 

if mounter.canSafelySkipMountPointCheck() {
    // UnmountwithForce()
    // removePath if no error
} else {
   // the old logic
}

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (jingxu97) [inline staging/src/k8s.io/mount-utils/mount_helper_common.go:93]:
```
if mounter.canSafelySkipMountPointCheck() || corruptedMnt {
  // mounter.Unmount
  // removePath if no error from unmount
} else {
  // removePathIfNotMountPoint
  // mounter.Unmount
  // removePathIfNotMountPoint if no error
}
```

Comment (jingxu97) [reply staging/src/k8s.io/mount-utils/mount_helper_common.go:93]: in previous logic, if unmount returns no error, we still check mountpoint before removing path. Is this really necessary? If umount command returns with no error, does it indicate the path is no longer a mount point? We should just call removePath instead of removePathIfNotMountPoint, right?

Comment (mkimuram) [reply staging/src/k8s.io/mount-utils/mount_helper_common.go:93]: I think that the previous discussion from [here](https://github.com/kubernetes/kubernetes/pull/109117#issuecomment-1092163588) is related to this question. If we can assume that we can rely on rm command's failure to check if there is something mounted, it would be true.

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (jingxu97) [inline staging/src/k8s.io/mount-utils/mount_helper_common.go:108]: actually in line 103, after your change, how can we make sure it is a mountpoint? as long as mounter.canSafelySkipMountPointCheck() is true, if mountPath is not a mount point, it will return err at line 105, right?

File: staging/src/k8s.io/mount-utils/mount_linux.go
Comment (cartermckinnon) [reply staging/src/k8s.io/mount-utils/mount_linux.go:74]: It uses `libmount`, which has heuristics to avoid a mtab check whenever possible. It's implementation of a mtab check is also more efficient than what's happening in this library. Relevant function: https://github.com/util-linux/util-linux/blob/master/libmount/src/context_umount.c#L47

Another aspect of this PR is removing redundant checks. Even if the mount point check in this library had the same performance as `umount`'s, part of the issue is that the check is happening twice.

File: staging/src/k8s.io/mount-utils/mount_linux.go
Comment (jingxu97) [reply staging/src/k8s.io/mount-utils/mount_linux.go:242]: maybe make it level 4?

Also in the below func (mounter *Mounter) Unmount(target string), it does not try to lookPath for umount, so is it needed here?

File: staging/src/k8s.io/mount-utils/mount_linux.go
Comment (jingxu97) [reply staging/src/k8s.io/mount-utils/mount_linux.go:345]: put a log here, and update Unmount comment. 

Since we are changing the library here, we also need to make sure this behavior change will not affect any other logic.

File: staging/src/k8s.io/mount-utils/mount_linux.go
Comment (cartermckinnon) [reply staging/src/k8s.io/mount-utils/mount_linux.go]: I've changed these to V(4) and removed the LookPath.

**Comment by cartermckinnon** [comment]
@jingxu97 Can you take another look at this?

File: staging/src/k8s.io/mount-utils/mount_linux.go
Comment (jingxu97) [reply staging/src/k8s.io/mount-utils/mount_linux.go:342]: many in-tree drivers uses this function, could you double check this behavior change will not affect any of them?

**Comment by cartermckinnon** [comment]
The only notable change in behavior I see is:

**Before this PR**, we may try to delete a target that is still a mount point, raising an error. This would happen if the wrong `extensiveMountPointCheck` value is used (`false` instead of `true`). **After this PR**, we will `Unmount` the target before attempting to delete it, resulting in expected behavior -- the mount will be successfully cleaned up and no error will be returned.

I can't find any usage of `CleanupMountPoint`/`CleanupMountWithForce` that is relying on a specific error being returned; I don't see any concerns around silencing the `not mounted` error.

For future reference, the only in-tree packages using the extensive mount point check (consistent read) for `CleanupMountPoint` are:
- `pkg/volume/nfs`
- `pkg/volume/local`
- `pkg/volume/util/subpath`

And the only package using `CleanupMountWithForce` is `pkg/volume/nfs`.

File: staging/src/k8s.io/mount-utils/mount_linux.go
Comment (cartermckinnon) [reply staging/src/k8s.io/mount-utils/mount_linux.go:342]: From a quick search, I think corrupted mounts often result in `failed: Operation not permitted` from `umount`. That behavior won't change with this PR, the error will be returned.

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (jingxu97) [inline staging/src/k8s.io/mount-utils/mount_helper_common.go:56]: maybe in the log, also show the value of corruptedMnt and canSafelySkipMountPointCheck, so we understand what is the situation there? the same applied the following change.

**Comment by jingxu97** [comment]
> The only notable change in behavior I see is:
> 
> **Before this PR**, we may try to delete a target that is still a mount point, raising an error. This would happen if the wrong `extensiveMountPointCheck` value is used (`false` instead of `true`). **After this PR**, we will `Unmount` the target before attempting to delete it, resulting in expected behavior -- the mount will be successfully cleaned up and no error will be returned.
> 
> I can't find any usage of `CleanupMountPoint`/`CleanupMountWithForce` that is relying on a specific error being returned; I don't see any concerns around silencing the `not mounted` error.
> 
> For future reference, the only in-tree packages using the extensive mount point check (consistent read) for `CleanupMountPoint` are:
> 
> * `pkg/volume/nfs`
> * `pkg/volume/local`
> * `pkg/volume/util/subpath`
> 
> And the only package using `CleanupMountWithForce` is `pkg/volume/nfs`.

Could you also include this behavior change in release note?

File: staging/src/k8s.io/mount-utils/mount_helper_common.go
Comment (cartermckinnon) [reply staging/src/k8s.io/mount-utils/mount_helper_common.go]: Sure, added in the latest commit.

**Comment by cartermckinnon** [comment]
@jingxu97 I've updated the release note, let me know if it makes sense 😄

**Comment by jingxu97** [comment]
> inappropriately

The main thing I want to put in the release note is related to the umount function in mount util

> @jingxu97 I've updated the release note, let me know if it makes sense 😄

The main thing I want to mention in the release note is about Unmount function in mount-util. Before if it is not a mountpoint, it will return error. After this change, if skipmountpointcheck is true, Unmount will succeed.

```
Explicit (not sure this is accurate word) mount point checks will now be skipped during CleanupMountPoint/CleanupMountWithForce if "not a mount point" error could be returned when executing umount operation. Also note that before this change, an error would be returned when using Unmount function in mount util. After this change if skipmountpointcheck is true, Unmount will succeed if the path is not a mount point.
```

**Comment by cartermckinnon** [comment]
@jingxu97 edits look great to me, I think we're good here?

**Comment by jfarrell** [comment]
@jingxu97 any pending updates you are waiting on for this for final review/sign-off or are we all set to be able to merge this?

**Comment by jingxu97** [comment]
Thank you very much for the great effort on this PR! Sorry for the delayed review.

**Pull request merged**
